src/lib/chart/dendrogram.typ

```diff
@@ -0,0 +1,312 @@
+)
+
+// Valid dendrogram modes
+#let dendrogram-modes = (
+  "vertical",
+  "horizontal"
+)
+
+// Functions for max value calculation
+#let dendrogram-max-value-fn = (
+  vertical: (data, value-key) => {
+    calc.max(0, ..data.map(t => t.at(value-key)))
+  },
+  horizontal: (data, value-key) => {
+    calc.max(0, ..data.map(t => t.at(value-key)))
+  },
+)
+
+// Functions for min value calculation
+#let dendrogram-min-value-fn = (
+  vertical: (data, value-key) => {
+    calc.min(0, ..data.map(t => t.at(value-key)))
+  },
+  horizontal: (data, value-key) => {
+    calc.min(0, ..data.map(t => t.at(value-key)))
+  },
+)
+
+/// Draw a dendrogram. A dendrogram is a chat that relative distances higher
+/// dimensional spaces. It is often used by data scientists in clustering
+/// analyses.
+///
+/// *Style root*: `dendrogram`.
+///
+/// - data (array): Array of data rows, with each entry representing a leaf on
+///                 the dendrogram. A row can be of type array or dictionary,
+///                 with `x1-key`, `x2-key`, and `height-key` being the keys
+///                 used to axcess a row's links and heights.
+///
+///                 *Example*
+///                 ```typc
+///                 ((1, 2, 0.5), (3, 4, 1), (6, 7, 2), (5, 8, 2.5))
+///                 ```
+/// - x1-key (int,string): Key to access the first cluster of a data row. This 
+///                        key is used as argument to the rows `.at(..)` 
+///                        function.
+/// - x2-key (int,string): Key to access the second cluster of a data row. 
+///                        This key is used as argument to the rows `.at(..)` 
+///                        function.
+/// - value-key (int,string): Key(s) to access value(s) of data row.
+///                           These keys are used as argument to the
+///                           rows `.at(..)` function.
+/// - mode (string): Chart mode:
+///                  - `"vertical"` -- Vertically displayed dendrogram
+/// - size (array): Chart size as width and height tuple in canvas units;
+///                 height can be set to `auto`.
+/// - line-style (style,function): Style or function (idx => style) to use for
+///                               each leaf, accepts a palette function.
+/// - x-label (content,none): X axis label
+/// - y-tick-step (float): Step size of y axis ticks 
+/// - x-ticks (array): List of tick values or value/label tuples
+///
+///                    *Example*
+///                    
+///                    `(1, 5, 10)` or `((1, [One]), (2, [Two]), (10, [Ten]))`
+/// - y-ticks (array): List of tick values or value/label tuples
+///
+///                    *Example*
+///                    
+///                    `(1, 5, 10)` or `((1, [One]), (2, [Two]), (10, [Ten]))`
+/// - y-unit (content,auto): Tick suffix added to each tick label
+/// - y-label (content,none): Y axis label
+/// - y-min (number,auto): Y axis minimum value
+/// - y-max (number,auto): Y axis maximum value
```

Comment:
what happens when they are `auto`?

Reply:
I haven't actually covered the cases where the given y range is smaller than that which is being rendered. It's probably a better idea to take these options away, as changing them to anything other than auto will never give you as nice a plot?

Reply:
I don't have much of an opinion but its probably best to at least let the user have the option to change it just in case.